src/client.rs

```diff
@@ -31,25 +31,23 @@ use crate::{
 /// for concurrent mutation. Under the hood, it corresponds to a
 /// Box<ClientConfig>.
 /// https://docs.rs/rustls/0.20.0/rustls/struct.ClientConfig.html
-pub struct rustls_client_config_builder_wants_verifier {
+pub struct rustls_client_config_builder {
     // We use the opaque struct pattern to tell C about our types without
     // telling them what's inside.
     // https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs
     _private: [u8; 0],
 }
 
-impl CastPtr for rustls_client_config_builder_wants_verifier {
-    type RustType = ConfigBuilder<ClientConfig, WantsVerifier>;
-}
-
-impl BoxCastPtr for rustls_client_config_builder_wants_verifier {}
-
-pub struct rustls_client_config_builder {
-    _private: [u8; 0],
+pub(crate) struct ClientConfigBuilder {
+    base: rustls::ConfigBuilder<ClientConfig, WantsVerifier>,
+    verifier: Arc<dyn ServerCertVerifier>,
```

Comment:
nit: instead of providing `NoneVerifier` to act as the default/sentinel value indicating that the client hasn't set a `ServerCertVerifier`, why not use `Option<Arc<dyn ServerCertVerifier>>`, analogously to how you handled `cert_resolver`?

Reply:
If I do that, then I have to be able to return an error from `rustls_client_config_builder_build` to indicate "hey you never told me how to verify server certificates." That's a possibility, and there's an argument that in the builder pattern, we should defer _all_ errors until the build step. But for now I wanted to hew somewhat to the existing 0.7.0 behavior which is "if you forget to set root certs or a custom verifier, you will get errors on every connection."

Reply:
Correct. `rustls` requires you to either set root certificates or set a custom verifier: https://docs.rs/rustls/0.20.0/rustls/struct.ConfigBuilder.html#impl-3.